File: app/analysis/trend_analyzer.py
"""
Trend analyzer — computes month-over-month changes and linear trend direction
for key financial metrics.
"""

from dataclasses import dataclass, field
from typing import Dict, List, Optional, Tuple

from app.models.statement import FinancialStatement


@dataclass
class TrendSeries:
    metric: str
    label: str
    months: List[str]
    values: List[Optional[float]]
    mom_changes: List[Optional[float]]    # month-over-month absolute change
    mom_pct_changes: List[Optional[float]]  # month-over-month % change
    trend_direction: str                  # "improving" | "worsening" | "stable" | "volatile"
    overall_pct_change: Optional[float]   # first→last % change
    avg_value: Optional[float]
    peak_month: Optional[str]
    trough_month: Optional[str]


@dataclass
class TrendReport:
    series: Dict[str, TrendSeries] = field(default_factory=dict)

    def get(self, name: str) -> Optional[TrendSeries]:
        return self.series.get(name)


# Metrics to track, and whether "increasing" is good (True) or bad (False)
TRACKED_METRICS: List[Tuple[str, str, bool]] = [
    ("total_revenue",            "Total Revenue",            True),
    ("total_operating_expenses", "Total Operating Expenses", False),
    ("noi",                      "Net Operating Income",     True),
    ("vacancy_loss",             "Vacancy Loss",             False),
    ("controllable_expenses",    "Controllable Expenses",    False),
    ("non_controllable_expenses","Non-Controllable Expenses",False),
    ("total_payroll",            "Total Payroll",            False),
    ("management_fees",          "Management Fees",          False),
    ("cash_flow",                "Cash Flow",                True),
    ("net_income",               "Net Income",               True),
]


def analyze_trends(stmt: FinancialStatement) -> TrendReport:
    report = TrendReport()

    for metric_key, label, higher_is_better in TRACKED_METRICS:
        item = stmt.get_figure(metric_key)
        if item is None:
            continue

        values = [item.monthly_values.get(m) for m in stmt.months]
        numeric = [(i, v) for i, v in enumerate(values) if v is not None]

        if len(numeric) < 2:
            continue

        # Month-over-month changes
        mom_abs: List[Optional[float]] = [None]
        mom_pct: List[Optional[float]] = [None]
        for i in range(1, len(values)):
            prev = values[i - 1]
            curr = values[i]
            if prev is None or curr is None:
                mom_abs.append(None)
                mom_pct.append(None)
            else:
                mom_abs.append(curr - prev)
                mom_pct.append(((curr - prev) / abs(prev)) * 100 if prev != 0 else None)

        # Overall trend
        first_val = values[numeric[0][0]]
        last_val  = values[numeric[-1][0]]
        overall_pct = ((last_val - first_val) / abs(first_val) * 100) if first_val and first_val != 0 else None

        # Average
        nums = [v for v in values if v is not None]
        avg = sum(nums) / len(nums) if nums else None

        # Peak / trough
        peak_idx   = max(numeric, key=lambda x: x[1])[0]
        trough_idx = min(numeric, key=lambda x: x[1])[0]

        # Trend direction via linear slope
        direction = _classify_trend(numeric, higher_is_better, mom_pct)

        report.series[metric_key] = TrendSeries(
            metric=metric_key,
            label=label,
            months=stmt.months,
            values=values,
            mom_changes=mom_abs,
            mom_pct_changes=mom_pct,
            trend_direction=direction,
            overall_pct_change=overall_pct,
            avg_value=avg,
            peak_month=stmt.months[peak_idx] if nums else None,
            trough_month=stmt.months[trough_idx] if nums else None,
        )

    return report


def _classify_trend(
    numeric: List[Tuple[int, float]],
    higher_is_better: bool,
    mom_pct: List[Optional[float]],
) -> str:
    if len(numeric) < 3:
        return "stable"

    indices = [i for i, _ in numeric]
    vals    = [v for _, v in numeric]

    # Simple linear regression slope
    n = len(indices)
    mean_x = sum(indices) / n
    mean_y = sum(vals) / n
    num = sum((x - mean_x) * (y - mean_y) for x, y in zip(indices, vals))
    den = sum((x - mean_x) ** 2 for x in indices)
    slope = num / den if den != 0 else 0

    # Volatility: std dev of mom_pct changes
    valid_pcts = [p for p in mom_pct if p is not None]
    if len(valid_pcts) >= 3:
        import statistics
        try:
            stdev = statistics.stdev(valid_pcts)
            if stdev > 20:
                return "volatile"
        except Exception:
            pass

    # Classify slope relative to mean value magnitude
    mean_val = abs(mean_y) if mean_y != 0 else 1
    slope_pct = (slope / mean_val) * 100

    if abs(slope_pct) < 1.0:
        return "stable"

    if slope > 0:
        return "improving" if higher_is_better else "worsening"
    else:
        return "worsening" if higher_is_better else "improving"

File: lib/analysis/trend-analyzer.ts
import type { FinancialStatement, TrendReport, TrendSeries } from '../models/statement';

interface MetricConfig {
  key: string;
  label: string;
  higherIsBetter: boolean;
}

const TRACKED_METRICS: MetricConfig[] = [
  { key: 'total_revenue', label: 'Total Revenue', higherIsBetter: true },
  { key: 'total_operating_expenses', label: 'Total Operating Expenses', higherIsBetter: false },
  { key: 'noi', label: 'Net Operating Income', higherIsBetter: true },
  { key: 'vacancy_loss', label: 'Vacancy Loss', higherIsBetter: false },
  { key: 'controllable_expenses', label: 'Controllable Expenses', higherIsBetter: false },
  { key: 'non_controllable_expenses', label: 'Non-Controllable Expenses', higherIsBetter: false },
  { key: 'total_payroll', label: 'Total Payroll', higherIsBetter: false },
  { key: 'management_fees', label: 'Management Fees', higherIsBetter: false },
  { key: 'cash_flow', label: 'Cash Flow', higherIsBetter: true },
  { key: 'net_income', label: 'Net Income', higherIsBetter: true },
];

function linearRegressionSlope(xVals: number[], yVals: number[]): number {
  const n = xVals.length;
  if (n < 2) return 0;
  const xMean = xVals.reduce((a, b) => a + b, 0) / n;
  const yMean = yVals.reduce((a, b) => a + b, 0) / n;
  const numerator = xVals.reduce((sum, x, i) => sum + (x - xMean) * (yVals[i] - yMean), 0);
  const denominator = xVals.reduce((sum, x) => sum + Math.pow(x - xMean, 2), 0);
  return denominator === 0 ? 0 : numerator / denominator;
}

function calcMean(values: number[]): number {
  if (values.length === 0) return 0;
  return values.reduce((a, b) => a + b, 0) / values.length;
}

function calcStdDev(values: number[], avg: number): number {
  if (values.length < 2) return 0;
  const variance = values.reduce((sum, v) => sum + Math.pow(v - avg, 2), 0) / values.length;
  return Math.sqrt(variance);
}

function determineTrendDirection(
  values: number[],
  months: string[],
  higherIsBetter: boolean,
): 'improving' | 'worsening' | 'stable' | 'volatile' {
  if (values.length < 2) return 'stable';

  const avg = calcMean(values);
  if (avg === 0) return 'stable';

  // Calculate MoM % changes for volatility
  const pctChanges: number[] = [];
  for (let i = 1; i < values.length; i++) {
    if (values[i - 1] !== 0) {
      pctChanges.push(((values[i] - values[i - 1]) / Math.abs(values[i - 1])) * 100);
    }
  }

  const xVals = months.map((_, i) => i);
  const slope = linearRegressionSlope(xVals, values);
  const slopePct = (slope / Math.abs(avg)) * 100;

  if (Math.abs(slopePct) < 1) return 'stable';

  if (pctChanges.length > 0) {
    const pctMean = calcMean(pctChanges);
    const pctStd = calcStdDev(pctChanges, pctMean);
    if (pctStd > 20) return 'volatile';
  }

  const isGrowing = slope > 0;
  if (higherIsBetter) {
    return isGrowing ? 'improving' : 'worsening';
  } else {
    return isGrowing ? 'worsening' : 'improving';
  }
}

export function analyzeTrends(statement: FinancialStatement): TrendReport {
  const { keyFigures, months } = statement;
  const series: TrendSeries[] = [];

  for (const { key, label, higherIsBetter } of TRACKED_METRICS) {
    const row = keyFigures[key];
    if (!row) continue;

    const values: Record<string, number | null> = {};
    for (const month of months) {
      values[month] = row.montlyValues[month] ?? null;
    }

    // MoM changes
    const momChanges: Record<string, number | null> = {};
    const momPctChanges: Record<string, number | null> = {};

    for (let i = 1; i < months.length; i++) {
      const prev = values[months[i - 1]];
      const curr = values[months[i]];
      if (prev !== null && curr !== null) {
        momChanges[months[i]] = curr - prev;
        momPctChanges[months[i]] = prev !== 0 ? ((curr - prev) / Math.abs(prev)) * 100 : null;
      } else {
        momChanges[months[i]] = null;
        momPctChanges[months[i]] = null;
      }
    }
    if (months.length > 0) {
      momChanges[months[0]] = null;
      momPctChanges[months[0]] = null;
    }

    // Filter non-null values for stats
    const nonNullMonths = months.filter(m => values[m] !== null);
    const nonNullVals = nonNullMonths.map(m => values[m] as number);

    const avgValue = nonNullVals.length > 0 ? calcMean(nonNullVals) : null;

    // Peak and trough
    let peakMonth: string | null = null;
    let troughMonth: string | null = null;
    if (nonNullVals.length > 0) {
      const maxVal = Math.max(...nonNullVals);
      const minVal = Math.min(...nonNullVals);
      peakMonth = nonNullMonths[nonNullVals.indexOf(maxVal)];
      troughMonth = nonNullMonths[nonNullVals.indexOf(minVal)];
    }

    // Overall % change (first to last non-null)
    let overallPctChange: number | null = null;
    if (nonNullVals.length >= 2) {
      const first = nonNullVals[0];
      const last = nonNullVals[nonNullVals.length - 1];
      overallPctChange = first !== 0 ? ((last - first) / Math.abs(first)) * 100 : null;
    }

    const trendDirection = nonNullVals.length >= 2
      ? determineTrendDirection(nonNullVals, nonNullMonths, higherIsBetter)
      : 'stable';

    series.push({
      metric: key,
      label,
      values,
      momChanges,
      momPctChanges,
      trendDirection,
      overallPctChange,
      peakMonth,
      troughMonth,
      avgValue,
    });
  }

  return { series };
}
